src/hpm.c

```diff
@@ -1,350 +1,356 @@
+        break;
+#else
+#define METAL_HPM_GET_COUNT_REG(x)                                             \
+    case METAL_HPM_COUNTER_##x:                                                \
+        __asm__ __volatile__("csrr %0, mhpmcounter" #x : "=r"(vl));            \
+        break;
+#endif
+
+/* Macro to clear HW performance monitor counter values */
+#if __riscv_xlen == 32
+#define METAL_HPM_CLR_COUNT_REG(x)                                             \
+    case METAL_HPM_COUNTER_##x:                                                \
+        __asm__ __volatile__("csrw mhpmcounter" #x "h, zero");                 \
+        __asm__ __volatile__("csrw mhpmcounter" #x ", zero");                  \
+        __asm__ __volatile__("csrw mhpmcounter" #x "h, zero");                 \
+        break;
+#else
+#define METAL_HPM_CLR_COUNT_REG(x)                                             \
+    case METAL_HPM_COUNTER_##x:                                                \
+        __asm__ __volatile__("csrw mhpmcounter" #x ", zero");                  \
+        break;
+#endif
+
+/* Max available counters */
+#define METAL_HPM_COUNT_MAX 32
+
+/* Macro to check for instruction trap */
+#define MCAUSE_ILLEGAL_INST 0x02
+
+/* Return codes */
+#define METAL_HPM_RET_OK 0
+#define METAL_HPM_RET_NOK 1
+
+int metal_hpm_init(struct metal_cpu *gcpu) {
+    struct __metal_driver_cpu *cpu = (void *)gcpu;
+
+    /* Check if counters are initialized or pointer is NULL */
+    if ((gcpu) && (cpu->hpm_count == 0)) {
+        metal_hpm_counter n;
+
+        /* Count number of available hardware performance counters */
+        cpu->hpm_count = METAL_HPM_COUNT_MAX;
+
+        /* mcycle, mtime and minstret counters are always available */
+        for (n = METAL_HPM_COUNTER_3; n < METAL_HPM_COUNTER_31; n++) {
+            metal_hpm_set_event(gcpu, n, 0xFFFFFFFF);
+
+            if (metal_hpm_get_event(gcpu, n) == 0) {
+                break;
+            }
+        }
+        cpu->hpm_count = n;
+
+        /* TODO: mcountinhibit csr is not yet accessible.
+         * As per latest RiscV privileged spec v1.11,
+         * mcountinhibit controls which of the counters increment.
+         * Unused counters can be disabled to reduce power consumption. */
+        /* Keep all counters disabled, enable them later on as needed. */
+        /* __asm__ __volatile__("csrw mcountinhibit, zero"); */
+
+        /* Clear all counters */
+        for (unsigned int i = 0; i < cpu->hpm_count; i++) {
+            metal_hpm_clr_event(gcpu, i, 0xFFFFFFFF);
+            metal_hpm_clear_counter(gcpu, i);
+        }
+    } else {
+        return METAL_HPM_RET_NOK;
+    }
+
+    return METAL_HPM_RET_OK;
+}
+
+int metal_hpm_disable(struct metal_cpu *gcpu) {
+    struct __metal_driver_cpu *cpu = (void *)gcpu;
+    uintptr_t temp, val = 0;
+
+    /* Check if pointer is NULL */
+    if (gcpu) {
+        /* Disable counter access */
+
+        __asm__ __volatile__("csrr %1, mtvec \n\t"
```

Comment:
This would not work, as asm does not guarantee across different asm lines. Please check with @jim-wilson 

What you had previously would work. You simply need to check if there is trap during in the init() and return NOT_OK
Application code should check for init() success before calling the reset of the HPM apis. Your the hpm example, simply exit with return code 0, for design that don't support HPM, ie init failed.

Reply:
The compiler doesn't guarantee that it won't optimize code around asms.  But the asms are volatile which limits the optimizations that can be performed.  And the first asm stores a value into temp that is used by the third asm.  It looks like it should work.

However, it can only work if it is optimized, and the optimizer manages to put temp into a register.  Without optimization, the local temp might be allocated to a stack slot, and then the compiler will emit an instruction to load temp into a register right before the third asm, which will fail if you take a trap and jump into the asm.  You can make it safer by requiring optimization somehow.  Or maybe by adding checks to ensure that temp is in a register, though it isn't immediately obvious how to do that.

Reply:
Combined multiple asm statements into one, to better avoid any unwanted side effects.